Classes/common/HTTP/CDTSessionCookieInterceptor.m

```diff
@@ -0,0 +1,147 @@
+    [request setValue:@"application/x-www-form-urlencoded" forHTTPHeaderField:@"Content-Type"];
+
+    dispatch_semaphore_t sema = dispatch_semaphore_create(0);
+    __block NSString *cookie = nil;
+    NSURLSessionDataTask *task = [self.urlSession
+        dataTaskWithRequest:request
+          completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
+            // all parameters to block are nullable, annotations removed
+            // to provide compatbility with xcode 6.4 and xcode 7
+            NSHTTPURLResponse *httpResp = (NSHTTPURLResponse *)response;
+            // handle response being nil
+            if ([httpResp statusCode] / 100 == 2) {
+                // we have the cookie. maybe
+                if (data && [self hasSessionStarted:data]) {
+                    // get the cookie from the header
+                    NSString *cookieHeader = [[httpResp allHeaderFields] valueForKey:@"Set-Cookie"];
+
+                    cookie = [cookieHeader componentsSeparatedByString:@";"][0];
+
+                } else {
+                    cookie = nil;
+                }
+            } else if ([httpResp statusCode] == 401) {
+                CDTLogError(CDTREPLICATION_LOG_CONTEXT, @"Credentials are incorrect, cookie "
+                                                        @"authentication will not be attempted "
+                                                        @"again by this interceptor object");
+                self.shouldMakeCookieRequest = NO;
+            } else if ([httpResp statusCode] / 100 == 5) {
+                CDTLogError(CDTREPLICATION_LOG_CONTEXT,
+                            @"Failed to get cookie from the server, response code was %ld cookie "
+                            @"authentication will not be attempted again by this interceptor "
+                            @"object",
+                            (long)[httpResp statusCode]);
+                self.shouldMakeCookieRequest = NO;
+            } else {
+                CDTLogError(CDTREPLICATION_LOG_CONTEXT,
+                            @"Failed to get cookie from the server,response code %ld cookie "
+                            @"authentication will not be attempted again by this interceptor "
+                            @"object",
+                            (long)[httpResp statusCode]);
+            }
+
+            dispatch_semaphore_signal(sema);
+
+          }];
+    [task resume];
+    dispatch_semaphore_wait(sema, dispatch_time(DISPATCH_TIME_NOW, 600 * NSEC_PER_SEC));
+
+    return cookie;
+}
+
+- (BOOL)hasSessionStarted:(nonnull NSData *)data
+{
+    NSDictionary *jsonResponse = [NSJSONSerialization JSONObjectWithData:data options:0 error:nil];
+
+    // only check for ok:true, https://issues.apache.org/jira/browse/COUCHDB-1356
+    // means we cannot check that the name returned is the one we sent.
+    return
+        [jsonResponse objectForKey:@"ok"] != nil && [[jsonResponse objectForKey:@"ok"] boolValue];
```

Comment:
You can use `jsonResponse[@"ok"]`. If you shorted the var name to `json` you'll be able to fit everything on one line.

In particular, I think `[nil boolValue]` is `NO` -- double check, but you might not need the guarding clause.